### PR TITLE
feat(menu): adds x offset and y offset properties to menu

### PR DESCRIPTION
**Pull request by shaunak1111**
adds offset to the menu from the trigger button, throws error when a invalid number is supplied

Closes: #1672

### CODE REVIEW

**Comment by shaunak1111** [comment]
@kara @jelbourn please review PR. 
Added x and y offset property to menu

**Comment by willshowell** [comment]
Why not switch the approach to storing the value as a number? It would reduce the conversions, and if the input value is a string, you could just cast it as a number in the setter.

**Comment by shaunak1111** [comment]
@willshowell Yeah sure we can switch to number will give it a shot

**Comment by shaunak1111** [comment]
@willshowell 
I want to detect the case when user enters a empty string i.e ''. Casting empty string  to number results in value 0, therefore such conversions are needed and this approach has been taken.
[https://stackoverflow.com/questions/30199707/how-to-type-cast-from-string-which-is-originally-number-to-number](string to number casting)

**Comment by willshowell** [comment]
@shaunak1111 

I don't really see the issue with using the `+value || 0` approach. If the user (inappropriately) tries `yOffset=""` or `yOffset="clown"`, then the menu won't be offset and their error will be quickly diagnosed.

The user really should be binding as a value instead of a string, and I don't think I agree that casting back and forth several times just to catch invalid strings is the best approach.
```
<md-menu [xOffset]="myValue" [yOffset]="-15" ... >
```

In any case, thanks for the PR and hopefully someone from the team will review/comment soon! 😀

**Comment by shaunak1111** [comment]
@willshowell
All I wanted is to throw an error when user does something <md-menu [xOffset]="" [yOffset]="-15" >
Anyways I also think casting back and forth unnecessarily complicates the code, +value || 0 /  Number(value) || 0 is a good approach. Let's wait for the comments and I will change accordingly, anyways it is a minor change. 😀

Thanks

File: src/lib/menu/menu-directive.ts [hunk 1]
```diff
@@ -46,6 +51,8 @@ export class MdMenu implements AfterContentInit, MdMenuPanel, OnDestroy {
   private _keyManager: FocusKeyManager;
   private _xPosition: MenuPositionX = 'after';
   private _yPosition: MenuPositionY = 'below';
+  private _xOffset: string = '';
+  private _yOffset: string = '';
```
Comment (kara) [inline src/lib/menu/menu-directive.ts]: Prefer these to be numbers

File: src/lib/menu/menu.spec.ts [hunk 2]
```diff
@@ -204,10 +211,47 @@ describe('MdMenu', () => {
...
+      const overlayRectBottomOffset = overlayRect.top -  triggerRect.top;
+
+      expect(Math.floor(overlayRectLeftOffset))
+        .toBe(Math.floor(xOffsetToSet),
+          `Expected x offset of the menu to be set`);
+
+      expect(Math.floor(overlayRectBottomOffset))
+        .toBe(Math.floor(yOffsetToSet),
+          `Expected y offset of the menu to be set`);
+    });
```
Comment (shaunak1111) [inline src/lib/menu/menu.spec.ts]: Sure @kara I will write test case using number.
Thanks for the review

**Comment by shaunak1111** [comment]
@kara I have made the necessary changes as suggested. Please review the PR.
Thanks

File: src/lib/menu/menu.spec.ts [hunk 3]
```diff
@@ -1043,7 +1082,7 @@ class SimpleMenu {
 @Component({
   template: `
     <button [mdMenuTriggerFor]="menu" #triggerEl>Toggle menu</button>
-    <md-menu [xPosition]="xPosition" [yPosition]="yPosition" #menu="mdMenu">
+    <md-menu [xPosition]="xPosition" [yPosition]="yPosition"#menu="mdMenu">
```
Comment (willshowell) [inline src/lib/menu/menu.spec.ts]: Missing space before `#menu`

File: src/lib/menu/menu-directive.ts [hunk 4]
```diff
@@ -110,6 +118,28 @@ export class MdMenu implements AfterContentInit, MdMenuPanel, OnDestroy {
     this.setPositionClasses();
   }
 
+  /** Amount of x offset of the menu. */
```
Comment (willshowell) [inline src/lib/menu/menu-directive.ts]: Should this mention units?

File: src/lib/menu/menu.spec.ts (hunk 3 above)
Comment (shaunak1111) [reply src/lib/menu/menu.spec.ts]: @willshowell  I will make the changes and raise the PR

**Comment by shaunak1111** [comment]
@kara I have made the changes recommended by you. Please review the PR.
@willshowell I have incorporated your suggested changes.
Thanks

**Comment by shaunak1111** [comment]
@kara I have raised the PR with latest necessary changes suggested

Thanks

**Comment by shaunak1111** [comment]
@kara @willshowell 

Raised a PR again to resolve the conflicts. Please review the PR.

File: src/lib/menu/menu-directive.ts [hunk 5]
```diff
@@ -114,6 +123,28 @@ export class MatMenu implements AfterContentInit, MatMenuPanel, OnDestroy {
     this.setPositionClasses();
   }
 
+  /** Units in pixels to be set in x offset of the menu. */
+  @Input()
+  get xOffset() { return this._xOffset; }
+  set xOffset(value: number) {
+    let offset = +value;
+    if (isNaN(offset)) {
```
Comment (willshowell) [inline src/lib/menu/menu-directive.ts]: Maybe use [`coerceNumberProperty `](https://github.com/angular/material2/blob/master/src/cdk/coercion/number-property.ts)? I guess it depends on how important throwing an error is.

File: src/lib/menu/menu-directive.ts [hunk 6]
```diff
@@ -71,11 +76,15 @@ const MAT_MENU_BASE_ELEVATION = 2;
   ],
   exportAs: 'matMenu'
 })
+
+
```
Comment (willshowell) [inline src/lib/menu/menu-directive.ts]: Extra newlines

File: src/lib/menu/menu-trigger.ts [hunk 7]
```diff
@@ -332,6 +332,7 @@ export class MatMenuTrigger implements AfterContentInit, OnDestroy {
    * to the trigger.
    * @returns ConnectedPositionStrategy
    */
+
```
Comment (willshowell) [inline src/lib/menu/menu-trigger.ts]: Extra newline

**Comment by jmichaelprobert** [comment]
Has this been abandoned? (Would really love to see this implemented)

**Comment by jelbourn** [comment]
Closing this since it's become quite stale. Feel free to open a new PR if this is still relevant.

**Pull request closed without merging**